src/jdk.jdi/share/classes/com/sun/tools/jdi/StackFrameImpl.java

```diff
@@ -395,29 +395,25 @@ public PacketStream send() {
         } catch (JDWPException exc) {
             switch (exc.errorCode()) {
             case JDWP.Error.OPAQUE_FRAME:
-                if (thread.isVirtual()) {
-                    // We first need to find out if the current frame is native, or if the
-                    // previous frame is native, in which case we throw NativeMethodException
-                    for (int i = 0; i < 2; i++) {
-                        StackFrameImpl sf;
-                        try {
-                            sf = (StackFrameImpl)thread.frame(i);
-                        } catch (IndexOutOfBoundsException e) {
-                            // This should never happen, but we need to check for it.
-                            break;
-                        }
-                        sf.validateStackFrame();
-                        MethodImpl meth = (MethodImpl)sf.location().method();
-                        if (meth.isNative()) {
-                            throw new NativeMethodException();
-                        }
+                // We first need to find out if the current frame is native, or if the
+                // previous frame is native, in which case we throw NativeMethodException
+                for (int i = 0; i < 2; i++) {
+                    StackFrameImpl sf;
```

Comment:
There is nothing implementation-specific here.
I'd suggest to:
- `StackFrameImpl` -> `StackFrame`;
- `MethodImpl` -> `Method`;
- remove `validateStackFrame` at line 408 ('MethodImpl.location()' calls it)